File: assets/Scripts/Camera/CameraZoomController.js
// FILE: Scripts/Camera/CameraZoomController.js
"use strict";

/**
 * CameraZoomController
 * - Progressive zoom steps (e.g. 2,4,8,16,32)
 * - Smooth interpolation current -> target
 * - Input: INP.consumeWheelDelta() (global) + optional ctx.input.zoomIn/zoomOut
 *
 * Usage:
 *   const zoom = new CameraZoomController({ steps:[2,4,8,16,32], index:2 });
 *   zoom.update(dt, { input:{ zoomIn:false, zoomOut:false } });
 *   const dist = zoom.value();
 */

function clamp(v, lo, hi) { return Math.max(lo, Math.min(hi, v)); }
function num(v, fb) { const n = +v; return Number.isFinite(n) ? n : (fb || 0); }

function expSmooth(cur, target, smooth, dt) {
    const s = Math.max(0, num(smooth, 0));
    if (s <= 0) return target;
    const a = 1 - Math.exp(-s * dt);
    return cur + (target - cur) * a;
}

class CameraZoomController {
    constructor(cfg) {
        cfg = cfg || {};

        this.steps = Array.isArray(cfg.steps) && cfg.steps.length ? cfg.steps.slice() : [2, 4, 8, 16, 32];

        this.minIndex = (typeof cfg.minIndex === "number") ? (cfg.minIndex | 0) : 0;
        this.maxIndex = (typeof cfg.maxIndex === "number") ? (cfg.maxIndex | 0) : (this.steps.length - 1);

        this.minIndex = clamp(this.minIndex, 0, this.steps.length - 1) | 0;
        this.maxIndex = clamp(this.maxIndex, this.minIndex, this.steps.length - 1) | 0;

        this.index = (typeof cfg.index === "number") ? (cfg.index | 0) : clamp(((this.steps.length / 2) | 0), this.minIndex, this.maxIndex);
        this.index = clamp(this.index, this.minIndex, this.maxIndex) | 0;

        this.target = num(cfg.target, num(this.steps[this.index], 8));
        this.current = num(cfg.current, this.target);

        this.smooth = num(cfg.smooth, 18.0);
        this.cooldown = num(cfg.cooldown, 0.08);
        this._cd = 0;

        // Wheel direction: true -> invert
        this.invertWheel = !!cfg.invertWheel;

        // Optional hard clamp (in case you put weird steps)
        this.min = num(cfg.min, 1.2);
        this.max = num(cfg.max, 120.0);

        // How many steps to move per tick (normally 1)
        this.stepStride = Math.max(1, (cfg.stepStride | 0) || 1);
    }

    configure(cfg) {
        if (!cfg) return this;

        if (Array.isArray(cfg.steps) && cfg.steps.length) {
            this.steps = cfg.steps.slice();
            this.minIndex = clamp(this.minIndex, 0, this.steps.length - 1) | 0;
            this.maxIndex = clamp(this.maxIndex, this.minIndex, this.steps.length - 1) | 0;
            this.index = clamp(this.index, this.minIndex, this.maxIndex) | 0;
            this.target = clamp(this.target, this.min, this.max);
            this.current = clamp(this.current, this.min, this.max);
        }

        if (typeof cfg.minIndex === "number") this.minIndex = clamp(cfg.minIndex | 0, 0, this.steps.length - 1) | 0;
        if (typeof cfg.maxIndex === "number") this.maxIndex = clamp(cfg.maxIndex | 0, this.minIndex, this.steps.length - 1) | 0;

        if (typeof cfg.index === "number") this.index = clamp(cfg.index | 0, this.minIndex, this.maxIndex) | 0;
        if (typeof cfg.smooth === "number") this.smooth = num(cfg.smooth, this.smooth);
        if (typeof cfg.cooldown === "number") this.cooldown = Math.max(0, num(cfg.cooldown, this.cooldown));
        if (typeof cfg.invertWheel === "boolean") this.invertWheel = cfg.invertWheel;

        if (typeof cfg.min === "number") this.min = num(cfg.min, this.min);
        if (typeof cfg.max === "number") this.max = Math.max(this.min, num(cfg.max, this.max));

        if (typeof cfg.stepStride === "number") this.stepStride = Math.max(1, cfg.stepStride | 0);

        // resync target to current index
        this.target = clamp(num(this.steps[this.index], this.target), this.min, this.max);
        this.current = clamp(this.current, this.min, this.max);

        return this;
    }

    reset(value) {
        if (typeof value === "number" && Number.isFinite(value)) {
            this.current = clamp(value, this.min, this.max);
            this.target = this.current;
        }
        this._cd = 0;
        return this;
    }

    value() { return this.current; }
    targetValue() { return this.target; }
    stepIndex() { return this.index; }

    setIndex(idx, snap) {
        const i = clamp((idx | 0), this.minIndex, this.maxIndex) | 0;
        this.index = i;
        this.target = clamp(num(this.steps[this.index], this.target), this.min, this.max);
        if (snap) this.current = this.target;
        return this;
    }

    _consumeWheel() {
        try {
            if (typeof INP !== "undefined" && INP && typeof INP.consumeWheelDelta === "function") {
                return num(INP.consumeWheelDelta(), 0);
            }
        } catch (_) {}
        return 0;
    }

    /**
     * update(dt, ctx)
     * ctx optional: { input:{zoomIn:boolean, zoomOut:boolean} }
     */
    update(dt, ctx) {
        dt = clamp(num(dt, 1 / 60), 0, 0.05);

        this._cd = Math.max(0, this._cd - dt);

        // read intent
        let want = 0;

        // wheel
        let w = this._consumeWheel();
        if (this.invertWheel) w = -w;
        if (w !== 0) want += (w > 0) ? 1 : -1;

        // optional key zoom
        const inp = ctx && ctx.input;
        if (inp) {
            if (inp.zoomIn) want += 1;
            if (inp.zoomOut) want -= 1;
        }

        if (want !== 0 && this._cd <= 0) {
            // convention: want>0 means "zoom in" => smaller distance => index--
            const dir = (want > 0) ? -1 : 1;
            this.index = clamp(this.index + dir * this.stepStride, this.minIndex, this.maxIndex) | 0;
            this.target = clamp(num(this.steps[this.index], this.target), this.min, this.max);
            this._cd = this.cooldown;
        }

        this.current = expSmooth(this.current, this.target, this.smooth, dt);
        return this;
    }
}

module.exports = CameraZoomController;

File: modulesSrc/CameraModule/src/main/resources/kalitech/engine/modules/Camera/CameraZoomController.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaNumber = luaRuntime.LuaNumber
local LuaNumberIsFinite = luaRuntime.LuaNumberIsFinite
local LuaClass = luaRuntime.LuaClass
local LuaArrayIsArray = luaRuntime.LuaArrayIsArray
function clamp(self, v, lo, hi)
    local lua_temp_1
    if v < lo then
        lua_temp_1 = lo
    else
        local lua_temp_0
        if v > hi then
            lua_temp_0 = hi
        else
            lua_temp_0 = v
        end
        lua_temp_1 = lua_temp_0
    end
    return lua_temp_1
end
function num(self, v, fb)
    local n = LuaNumber(v)
    local lua_Number_isFinite_result_2
    if LuaNumberIsFinite(n) then
        lua_Number_isFinite_result_2 = n
    else
        lua_Number_isFinite_result_2 = fb
    end
    return lua_Number_isFinite_result_2
end
function expSmooth(self, cur, target, smooth, dt)
    local lua_temp_3
    if smooth > 0 then
        lua_temp_3 = smooth
    else
        lua_temp_3 = 0
    end
    local s = lua_temp_3
    if s == 0 then
        return target
    end
    local a = 1 - math.exp(LuaNumber(-s) * dt)
    return cur + (target - cur) * a
end
CameraZoomController = LuaClass()
CameraZoomController.name = "CameraZoomController"
function CameraZoomController.prototype.lua_constructor(self, cfg)
    cfg = cfg or KObject:create(nil)
    local lua_temp_4
    if LuaArrayIsArray(cfg.steps) and KLength(cfg.steps) > 0 then
        lua_temp_4 = cfg.steps
    else
        lua_temp_4 = {
            2,
            4,
            8,
            16,
            32
        }
    end
    local steps = lua_temp_4
    self.steps = KArrayOps.slice(steps)
    self.minIndex = bit32.bor(
        clamp(
            _G,
            bit32.bor(num(_G, cfg.minIndex, 0), 0),
            0,
            KLength(self.steps) - 1
        ),
        0
    )
    local lua_temp_5
    if cfg.maxIndex ~= nil then
        lua_temp_5 = bit32.bor(num(_G, cfg.maxIndex, KLength(self.steps) - 1), 0)
    else
        lua_temp_5 = KLength(self.steps) - 1
    end
    self.maxIndex = lua_temp_5
    self.maxIndex = bit32.bor(
        clamp(_G, self.maxIndex, self.minIndex, KLength(self.steps) - 1),
        0
    )
    local mid = bit32.bor(KLength(self.steps) / 2, 0)
    local lua_temp_6
    if cfg.index ~= nil then
        lua_temp_6 = bit32.bor(num(_G, cfg.index, mid), 0)
    else
        lua_temp_6 = mid
    end
    self.index = lua_temp_6
    self.index = bit32.bor(
        clamp(_G, self.index, self.minIndex, self.maxIndex),
        0
    )
    self.min = num(_G, cfg.min, 1.2)
    self.max = math.max(
        self.min,
        num(_G, cfg.max, 120)
    )
    self.target = clamp(
        _G,
        num(
            _G,
            cfg.target,
            num(_G, KIndex(self.steps, self.index), 8)
        ),
        self.min,
        self.max
    )
    self.current = clamp(
        _G,
        num(_G, cfg.current, self.target),
        self.min,
        self.max
    )
    self.smooth = num(_G, cfg.smooth, 18)
    self.cooldown = math.max(
        0,
        num(_G, cfg.cooldown, 0.08)
    )
    self._cd = 0
    self.invertWheel = not not cfg.invertWheel
    self.stepStride = math.max(
        1,
        bit32.bor(num(_G, cfg.stepStride, 1), 0)
    )
end
function CameraZoomController.prototype.configure(self, cfg)
    if not cfg then
        return self
    end
    if LuaArrayIsArray(cfg.steps) and KLength(cfg.steps) > 0 then
        self.steps = KArrayOps.slice(cfg.steps)
        self.minIndex = bit32.bor(
            clamp(_G, self.minIndex, 0, KLength(self.steps) - 1),
            0
        )
        self.maxIndex = bit32.bor(
            clamp(_G, self.maxIndex, self.minIndex, KLength(self.steps) - 1),
            0
        )
        self.index = bit32.bor(
            clamp(_G, self.index, self.minIndex, self.maxIndex),
            0
        )
    end
    if cfg.minIndex ~= nil then
        self.minIndex = bit32.bor(
            clamp(
                _G,
                bit32.bor(num(_G, cfg.minIndex, self.minIndex), 0),
                0,
                KLength(self.steps) - 1
            ),
            0
        )
    end
    if cfg.maxIndex ~= nil then
        self.maxIndex = bit32.bor(
            clamp(
                _G,
                bit32.bor(num(_G, cfg.maxIndex, self.maxIndex), 0),
                self.minIndex,
                KLength(self.steps) - 1
            ),
            0
        )
    end
    if cfg.index ~= nil then
        self.index = bit32.bor(
            clamp(
                _G,
                bit32.bor(num(_G, cfg.index, self.index), 0),
                self.minIndex,
                self.maxIndex
            ),
            0
        )
    end
    if cfg.smooth ~= nil then
        self.smooth = num(_G, cfg.smooth, self.smooth)
    end
    if cfg.cooldown ~= nil then
        self.cooldown = math.max(
            0,
            num(_G, cfg.cooldown, self.cooldown)
        )
    end
    if cfg.invertWheel ~= nil then
        self.invertWheel = not not cfg.invertWheel
    end
    if cfg.min ~= nil then
        self.min = num(_G, cfg.min, self.min)
    end
    if cfg.max ~= nil then
        self.max = math.max(
            self.min,
            num(_G, cfg.max, self.max)
        )
    end
    if cfg.stepStride ~= nil then
        self.stepStride = math.max(
            1,
            bit32.bor(num(_G, cfg.stepStride, self.stepStride), 0)
        )
    end
    self.target = clamp(
        _G,
        num(_G, KIndex(self.steps, self.index), self.target),
        self.min,
        self.max
    )
    self.current = clamp(_G, self.current, self.min, self.max)
    return self
end
function CameraZoomController.prototype.reset(self, value)
    if value ~= nil then
        local v = LuaNumber(value)
        if LuaNumberIsFinite(v) then
            self.current = clamp(_G, v, self.min, self.max)
            self.target = self.current
        end
    end
    self._cd = 0
    return self
end
function CameraZoomController.prototype.value(self)
    return self.current
end
function CameraZoomController.prototype.targetValue(self)
    return self.target
end
function CameraZoomController.prototype.stepIndex(self)
    return self.index
end
function CameraZoomController.prototype.setIndex(self, idx, snap)
    self.index = bit32.bor(
        clamp(
            _G,
            bit32.bor(idx, 0),
            self.minIndex,
            self.maxIndex
        ),
        0
    )
    self.target = clamp(
        _G,
        num(_G, KIndex(self.steps, self.index), self.target),
        self.min,
        self.max
    )
    if snap then
        self.current = self.target
    end
    return self
end
function CameraZoomController.prototype.update(self, dt, ctx)
    dt = clamp(
        _G,
        num(_G, dt, 1 / 60),
        0,
        0.05
    )
    self._cd = math.max(0, self._cd - dt)
    local want = 0
    local inp = ctx and ctx.input
    if inp then
        local w = num(_G, inp.wheel, 0)
        if w ~= 0 then
            local lua_table_invertWheel_7
            if self.invertWheel then
                lua_table_invertWheel_7 = LuaNumber(-w)
            else
                lua_table_invertWheel_7 = w
            end
            want = want + (lua_table_invertWheel_7 > 0 and 1 or -1)
        end
        if inp.zoomIn then
            want = want + 1
        end
        if inp.zoomOut then
            want = want - 1
        end
    end
    if want ~= 0 and self._cd == 0 then
        local dir = want > 0 and -1 or 1
        self.index = bit32.bor(
            clamp(_G, self.index + dir * self.stepStride, self.minIndex, self.maxIndex),
            0
        )
        self.target = clamp(
            _G,
            num(_G, KIndex(self.steps, self.index), self.target),
            self.min,
            self.max
        )
        self._cd = self.cooldown
    end
    self.current = expSmooth(
        _G,
        self.current,
        self.target,
        self.smooth,
        dt
    )
    return self
end
M = CameraZoomController

return M
